wallet-api/src/Wallet/API.hs

```diff
@@ -181,6 +183,19 @@ addresses = cata adr where
 newtype EventTrigger = EventTrigger { getEventTrigger :: Fix EventTriggerF }
     deriving (Eq, Ord, Show)
 
+-- | An action than can be run in response to a blockchain event. It receives
+--   the current block height, and the unspent transaction outputs that meet
+--   the condition specified in the trigger.
+newtype BlockchainAction m = BlockchainAction { runBlockchainAction :: Height -> AddressMap -> m () }
```

Comment:
Reading this it occurred to me that `Handler` might be a better term than `Action` because it's less generic and has a well-understood meaning that fits quite well here. What do you think?

> can't it then be a subsequent action that you have to do to go and look for unspent outputs

Right now there is no functionality in `WalletAPI` to do that (although it would be trivial to add). 

But I don't think it's a problem to have the handler type tightly coupled to the event. For example when a button was clicked you get the coordinates of the mouse etc.

And in our specific case, looking for the unspent outputs could also be quite difficult if you have a complex condition with `or` and `and` of different addresses. When the trigger fires you know that the condition holds, but you don't know which addresses made it true, and you might end up recreating the `checkTrigger` functionality in the handler. So I think it's better to give a constructive "proof" that the condition holds.

A more flexible solution would be to parameterise both trigger and handler types over some kind of primitive `p` (here, `p = Either FundsAtAddress BlockHeightRange`) and then allow different wallets with different `p`s. Maybe that could also be used to react to events from the outside world.

Reply:
Well, the problem is that if as now we realise that we want some additional bit of information after we get the notification then we have to go and change the type of the notification. Whereas if you just use an independent set of capabilities for getting data then it composes better.

> So I think it's better to give a constructive "proof" that the condition holds.

I do like this argument. Can we actually structure things in this way? At the moment you also can't distinguish the and/or case.

That said, I think this is compatible with the approach I described above - we should give you a proof tree of how we concluded your condition held, so you can tell which address it was that triggered it. But that doesn't require us to put the *actual outputs* in that proof tree. We can still require you to do that bit yourself.

Reply:
> That said, I think this is compatible with the approach I described above - we should give you a proof tree of how we concluded your condition held, so you can tell which address it was that triggered it. But that doesn't require us to put the actual outputs in that proof tree. We can still require you to do that bit yourself.

that sounds interesting, I'll see if I can come up with something nice